std-bits/base/src/main/java/org/enso/base/cache/LRUCache.java

```diff
@@ -0,0 +1,311 @@
+    }
+  }
+
+  /**
+   * IOExceptions thrown by the HTTP request are propagated; IOExceptions thrown while storing the
+   * data in the cache are caught.
+   */
+  private CacheResult<M> makeRequestAndCache(String cacheKey, ItemBuilder<M> itemBuilder)
+      throws IOException, InterruptedException, ResponseTooLargeException {
+    assert !cache.containsKey(cacheKey);
+
+    Item<M> item = itemBuilder.buildItem();
+
+    if (!item.shouldCache()) {
+      return new CacheResult<>(item.stream(), item.metadata());
+    }
+
+    if (item.sizeMaybe.isPresent()) {
+      long size = item.sizeMaybe().get();
+      if (size > getMaxFileSize()) {
+        throw new ResponseTooLargeException(getMaxFileSize());
+      }
+      makeRoomFor(size);
+    }
+
+    try {
+      // Download the response data.
+      File responseData = downloadResponseData(cacheKey, item);
+      M metadata = item.metadata();
+      long size = responseData.length();
+      ZonedDateTime expiry = getNow().plus(Duration.ofSeconds(item.ttl().get()));
+
+      // Create a cache entry.
+      var cacheEntry = new CacheEntry<>(responseData, metadata, size, expiry);
+      cache.put(cacheKey, cacheEntry);
+      markCacheEntryUsed(cacheKey);
+
+      // Clear out old entries to satisfy the total cache size limit. This might
+      // be necessary here if we didn't receive a correct content size value.
+      removeFilesToSatisfyLimit();
+
+      return getResultForCacheEntry(cacheKey);
+    } catch (IOException e) {
+      logger.log(
+          Level.WARNING, "Failure storing cache entry; will re-execute without caching: {}", e);
+      // Re-issue the request since we don't know if we've consumed any of the response.
+      Item<M> rerequested = itemBuilder.buildItem();
+      return new CacheResult<>(rerequested.stream(), rerequested.metadata());
+    }
+  }
+
+  /** Mark cache entry used and return a stream reading from the cache file. */
+  private CacheResult<M> getResultForCacheEntry(String cacheKey) throws IOException {
+    markCacheEntryUsed(cacheKey);
+    return new CacheResult<>(
+        new FileInputStream(cache.get(cacheKey).responseData), cache.get(cacheKey).metadata());
+  }
+
+  /**
+   * Read the repsonse data from the remote server into the cache file. If the downloaded data is
+   * over the file size limit, throw a ResponseTooLargeException.
+   */
+  private File downloadResponseData(String cacheKey, Item item)
+      throws IOException, ResponseTooLargeException {
+    File temp = File.createTempFile("LRUCache-" + cacheKey, "");
+    temp.deleteOnExit();
+    try {
+      var inputStream = item.stream();
+      try (var outputStream = new FileOutputStream(temp)) {
+
+        // Limit the download to getMaxFileSize().
+        boolean sizeOK = Stream_Utils.limitedCopy(inputStream, outputStream, getMaxFileSize());
+        if (!sizeOK) {
+          try {
+            temp.delete();
+            outputStream.close();
```

Comment:
The order of these operations seems suspicious.

Will the OS allow you to `delete` the file before closing the output stream _to that file_? Perhaps it's ok on Unix, as there it's okay to unlink opened files. But I imagine on Windows this will most likely fail.

I think you should also check if `temp.delete` returned `true`. I'd at least log a warning if it fails to delete.

Reply:
Done. I actually removed the call to `close` because it's all inside a try-with-resources block anyway.

Reply:
Hmmmm... But that means that the `delete` is still called inside of this try-with-resources block, meaning it's called when the stream is still open.

I'd like to make sure that this still correctly cleans up the file on Windows. From my experience with Windows I expect that it may likely be problematic.